### PR TITLE
Pointwise operator

### PR DESCRIPTION
**Pull request by nbouziani**
PointwiseOperator is the Firedrake bit of the ExternalOperator concept introduced in UFL (c.f. the associated pull request)

It handles the practical implementation thereof, especially it implements some of the subclasses covered by this concept : 

- PointexprOperator: Operator defined through an externally defined expression (e.g. lambda expression) passed in. This evaluation of this operator is done by merely evaluate this expression pointwise.

- PointsolveOperator:  Solve an externally defined algebraic relationship pointwise. For instance one can think to an implicit constitutive law that is part of a given PDE-based model.

There are more classes of function that stem from the ExternalOperator concept, but this pull request ONLY concern the architecture implemented to handle the AbstractPointwiseOperator (abstract base class) and 2 applications of it (the `point_solve` and the `point_expr`)

### CODE REVIEW

File: firedrake/pointwise_operators.py [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+from firedrake import utils, functionspaceimpl
+from pyop2.datatypes import ScalarType
+from ufl.log import error
+import sympy as sp
+import numpy as np
+from scipy import optimize
+
+
+class AbstractPointwiseOperator(Function, ExternalOperator, metaclass=ABCMeta):
+    r"""Asbtract base class from wich stem from all the firedrake practical implementation of the
```
Comment (dham) [inline firedrake/pointwise_operators.py]: Typo: Abstract

Comment (dham) [inline firedrake/pointwise_operators.py]: also "which"

Comment (dham) [reply firedrake/pointwise_operators.py]: the second "from" is spurious
Firedrake has a capital F
implementations (plural)

File: firedrake/pointwise_operators.py [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+            fs = ofs
+
+        Function.__init__(self, fs, val, name, dtype, count=self._count)  # count has been initialised in ExternalOperator.__init__
+
+        self._ufl_function_space = aa
+        self._original_function_space = fs
+        self.operator_data = operator_data
+
+        if extop_id is not None:
+            self.extop_id = extop_id
```
Comment (dham) [inline firedrake/pointwise_operators.py]: What does this do? I am a little scared of the caching implications of operator IDs.

File: firedrake/pointwise_operators.py [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+
+    @abstractmethod
+    def compute_derivatives(self):
+        """apply the derivatives on operator_data"""
+
+    @abstractmethod
+    def evaluate(self):
```
Comment (dham) [inline firedrake/pointwise_operators.py]: Random thought: do we do dependency analysis to make sure point wise operators get evaluated in the right order? Think about point wise operators with point wise operators in their arguments.

File: firedrake/pointwise_operators.py (hunk 2 above)
Comment (nbouziani) [reply firedrake/pointwise_operators.py]: It is a deprecated feature that I do not use anymore. I am wondering why I did not remove it, I will do it.

File: firedrake/pointwise_operators.py (hunk 3 above)
Comment (nbouziani) [reply firedrake/pointwise_operators.py]: I do not think we need to do it. The order in which the ExtOps get evaluated is based on the order on which they appear in the coefficients list of the form. And for external operator nesting, they will appear from bottom to top (cf. ufl/algorithms/analysis.py: extract_type function) so if I have two external operators p1 and p2 and a coefficient u and I consider the form p2(p1(u)), the coefficient list will be [u, p1, p2] and therefore p1 will be evaluated before p2.

File: firedrake/pointwise_operators.py (hunk 1 above)
Comment (nbouziani) [reply firedrake/pointwise_operators.py]: Oops, I changed that.

Comment (nbouziani) [reply firedrake/pointwise_operators.py]: Oops

**Comment by wence-** [comment]
Can this one be closed then?

**Pull request closed without merging**